### ISSUE TITLE
NumberKeypad: Design requirements

### ISSUE THREAD

**Issue by ariel-phet** [phetsims/scenery-phet#283]
@pixelzoom and @amanda-phet have pointed out the need for a general NumberKeypad design meeting. We should look into addressing the following questions as well as fleshing out and prioritizing other anticipated needs

- [ ] #271 
- [x] #272 
- [x] #275 
- [x] #280 
- [x] i18n requirements? (will be covered in #279 )
- [x] Scientific notation? (there is no requirement for this yet, so we've decided not to have any requirements along these lines and we will cross this bridge when we come to it)
- [x] layout options? (a vertically narrow, horizontally wide keypad might be advantageous in certain circumstances)

**Comment by ariel-phet** [comment]
@amanda-phet do you mind getting a design doc started?

**Comment by pixelzoom** [comment]
Also labeling for developer meeting, so we can sort out what needs to be done on NumberKeypad sooner vs later, who is lead developer on the design, who is doing the work, etc.

**Comment by pixelzoom** [comment]
11/10/16 design meeting notes:

• We need 2 separate keypads: one for numbers, one for entering equations
• Focus first on numbers, then on equations
• Currently missing from numbers keypad is sign (+/-) button (see #274)

**Comment by amanda-phet** [comment]
We should edit the bullets above to use "expressions", "strings", "text", or something along those lines, rather than "equations" (which implies that there is an equal sign being typed).

**Comment by pixelzoom** [comment]
"string" and "text" is way too general. How about "mathematical expressions"?

**Comment by pixelzoom** [comment]
"strings" and "text" is way too general, we're not designing a text editor. How about "mathematical expressions"?

**Comment by pixelzoom** [comment]
Revised 11/10/16 design meeting notes:

• We need 2 separate keypads: one for numbers, one for entering mathematical expressions
• Focus first on numbers, then on mathematical expressions
• Currently missing from numbers keypad is sign (+/-) button (see #274)

**Comment by jbphet** [comment]
@pixelzoom and @jonathanolson ,

@aadish and I (mostly Aadish) have done some work on the Keypad.  I think the current form captures most of the architecture that we discussed.  I’m mostly out for the next couple of weeks, but since this is going to be needed around the end of January, I thought it would be good for you to take a look and see if you think we’re generally on the right track.  There are a number of things that can be improved – for instance, there is a ‘value’ field in `AbstractKey` that can probably be removed. Please take a look at it from a more of a high level and give us your comments, and we will address them in the new year.

**Comment by jbphet** [comment]
The bulk of the code can be found in scenery-phet/keypad, and there is demo code in scenery-phet/demo.

**Comment by pixelzoom** [comment]
Comments on scenery-phet demo (prior to looking at code):

• When nothing has been entered on the keypad, the default "arithmetic value" should not be `0`.  That makes it impossible to distinguish between no entry and an explicit `0` entry without consulting the string value.  It would be preferable to use `null` to represent the absence of an arithmetic value.

• The demo itself is a little annoying, with stuff horizontally shifting around. This is probably just due to bad layout choices in the demo, but it's distracting, and potentially masks genuine problems. 

• In the demo, "Clear on next keypress" should be a checkbox, not a button.  It's a boolean state that's set, not an action that's performed. See the demo for NumberKeypad. (And if it doesn't already, the implementation must provide an API that allows the client to turn this on _and* off.)

• The demo doesn't demonstrate the decimal key.  It should, because it's essential to sims in development (e.g. unit-rates). Again, see the demo of NumberKeypad.

**Comment by pixelzoom** [comment]
My first pass at the code was a general review, look mainly at conformance to PhET guidelines. 

Issues:

- [ ] Keypad.createKeyNode - Private functions like this should be located _after_ the constructor.

- [ ] Keypad.createKeyNode has an `options` parameter (documented as optional), but has no `_.extend` call to define defaults. So `buttonFont` and `buttonColor` are not really options, they are effectively required parameters. Either add an `_.extend` call or function parameters. 

- [ ] Keypad and IntegerAccumulator constructor parameter `options` is not properly annotated as being optional. It should be `{Object} [options]`.

- [ ] We generally put `_.extend` as the first line in the constructor, since options are an important part of the public API.   This is currently not the case in Keypad and IntegerAccumulator, where `_.extend` follows the supertype constructor call.

- [ ] IntegerAccumulator  is missing visibility annotation for `validateAndProcessInput`.

- [ ] AbstractKey has insufficient documentation for constructor parameter `identifier`, and has no visibility annotations for properties defined in its constructor.

- [ ] AbstractKey `handleKeyPressed` is an abstract function. It should be annotated as `@abstract` and the default implementation should `throw Error`, not `assert`.   

- [ ] AbstractKey `handleKeyPressed` is missing the `keyAccumulator` parameter, which I only discovered by looking at subtype implementations.

- [ ] The documentation that is provided for AbstractKey `handleKeyPressed` is mostly redundant ("Function that is called by the key accumulator when this key is pressed, must be implemented in descendant classes.").  It's obviously a function, a descendent class (sic) is a subtype (or subclass, if you must), and the `@abstract` annotation would indicate that it needs to be implemented by subtypes.

- [ ] PlusMinusKey is missing `@public` and `@override` annotations for `handleKeyPressed`.

- [ ] Redundant documentation like "PlusMinusKey Class derived from AbstractKey class" is not very useful. The reader needs a quick overview of this type's role and responsibilities, not the specifics of the class (sic) hierarchy.

**Comment by pixelzoom** [comment]
I corrected many of the issues related to code guidelines (check off in the above check list).  I also added some `TODO` items in the code, where something still needs to be addressed. 

On to review of the code architecture...

**Comment by pixelzoom** [comment]
- [ ] Inappropriate property in AbstractKey base type.

The constructor signature of AbstractKey is:

```js
function AbstractKey( displayNode, value, identifier )
```

... where `value` is `{number}`.  Looking at non-numeric subtypes of AbstractKey (i.e. BackspaceKey, PlusMinusKey), I see that the `value` argument is `null`.  This is a fundamental problem - a numeric value has no business being in the base type, it is specific to keys that have a numeric value (i.e. DigitKey).

- [ ] BackspaceKey has required constructor parameters that could (should?) be options.

The constructor signature is:

```js
function BackspaceKey( width, height )
```

Why can't width and height be provided via options? There are surely reasonable defaults (as there apparently are for other keys).   Also wondering... Why does BackspaceKey have these options when Keypad is responsible for button size (via `buttonWidth` and `buttonHeight` options)? Why is Backspace key instantiated as `new BackspaceKey( DEFAULT_BUTTON_WIDTH, DEFAULT_BUTTON_HEIGHT )` in various Keypad layouts, when the button size might be changed via keypad options?



- [ ] Responsibility for key dimensions.

... and if BackspaceKey is responsible for its dimensions, then why aren't the other subtypes of AbstractKey also responsible for their dimensions?  Or the number of cells that they occupy in the grid? 

- [ ] Validation and processing of input

This is apparently done in AbstractKeyAccumulator subtypes via:

```js
validateAndProcessInput: function( accumulatedKeys )
```

I was surprised to see that this function has only one parameter, `accumulatedKeys`.  How does it "validate and process input" if the input is not provided?  Is this function mis-named? 

- [ ] Why isn't `_clearOnNextKeyPress` the responsibility of the base type?

IntegerAccumulator defines  `_clearOnNextKeyPress` and its associated setter/getter.  Why aren't these defined in AbstractKeyAccumulator?  I see nothing in the implementation to indicate that these are specific to IntegerAccumulator.

- [ ] Do you need a NumberAccumulator subtype?

There's a lot of stuff in IntegerAccumulator that would apply to (for example) DecimalAccumulator. Do you need to factor that out into a more general NumberAccumulator?

- [ ] Clients should not need to know about the accumulator.

In the scenery-phet demo, I see:

```js
555 var accumulator = new IntegerAccumulator( { maxLength: 5 } );
568  var keyPad = new Keypad( Keypad.PositiveAndNegativeIntegerLayout, accumulator
576 accumulator.clear();
583 accumulator.setClearOnNextKeyPress( true );
```

The accumulator should be an internal implementation detail, transparent to the client.  Clients shouldn't need to instantiate it, and certainly shouldn't need to explicitly clear it, etc.    These should be responsibilities of Keypad (or its subtypes).

- [ ] Absence of input should not be interpreted as `0`.

IntegerAccumulator.updateNumericalValue:

```js
return stringRepresentation.length > 0 ? parseInt( stringRepresentation, 10 ) : 0;
```

Absence of a value should most definitely _not_ be mapped to a value of `0`.  This makes it impossible for the client to determine whether `0` was actually entered without consulting the string value.  Recommended to change to:

```js
return ( stringRepresentation.length > 0 ) ? parseInt( stringRepresentation, 10 ) : null;
```

**Comment by pixelzoom** [comment]
- [ ] Specification of layouts should make it easy to see the keypad structure.

Specification of layouts is currently verbose, redundant, and makes it very difficult to see the structure of the keypad.  The current specification of only 2 layouts requires 145 lines!

Suggestions: 
• Use a 2-dimensional array instead of `row` and `column` fields. 
• Make `horizontalSpan` and `verticalSpan` properties of the keys, with default `1`.   

Then the specification of layouts would look like this:

```js
    PositiveIntegerLayout: [
      [ new DigitKey( 7 ), new DigitKey( 8 ), new DigitKey( 9 ) ],
      [ new DigitKey( 4 ), new DigitKey( 5 ), new DigitKey( 6 ) ],
      [ new DigitKey( 1 ), new DigitKey( 2 ), new DigitKey( 3 ) ],
      [ new DigitKey( 0, { horizontalSpan: 2 } ), new BackspaceKey() ]
    ],

    PositiveAndNegativeIntegerLayout: [
      [ new DigitKey( 7 ), new DigitKey( 8 ), new DigitKey( 9 ) ],
      [ new DigitKey( 4 ), new DigitKey( 5 ), new DigitKey( 6 ) ],
      [ new DigitKey( 1 ), new DigitKey( 2 ), new DigitKey( 3 ) ],
      [ new PlusMinusKey(), new DigitKey( 0 ), new BackspaceKey() ]
    ],
```

This makes it very easy to see the structure of the keypad, and which keys occupy more than 1 grid cell.

**Comment by pixelzoom** [comment]
- [ ] Use Unicode symbols for plus and minus.

In PlusMinusKey:

```js
20 AbstractKey.call( this, '+/-', null, 'PlusMinus' );
```

... should be:

```js
AbstractKey.call( this, '\u002b/\u2212', null, 'PlusMinus' );
```

**Comment by pixelzoom** [comment]
- [ ] Implement dispose

After addressing the issue of the accumulator being transparent to the client... You need to address the issue of disposing of a Keypad. In scenery-phet demo (ComponentsView), you have the client linking to Properties and there no `dispose` anywhere in the keypad implementation.

**Comment by pixelzoom** [comment]
- [ ] Why is a Property used for the array of accumulated keys?

In AbstractAccumulator:

```js
23    // @public - array property that tracks the accumulated key presses
24    this.accumulatedKeysProperty = new Property( [] );
```

I see no calls to `accumulatedKeysProperty.link`, so apparently nothing is observing it.  So why does it need to be a Property?  And if it does need to be observable, why is it not an ObservableArray?

**Comment by pixelzoom** [comment]
- [ ] IntegerAccumulator `updateStringValue` and `updateNumericalValue` are poorly named.

These functions don't actually update anything. They compute values, which are used elsewhere to update Properties.  More appropriate names would be `computeStringValue` and `computeNumericalValue`.

**Comment by pixelzoom** [comment]
- [ ] Redundant code in IntegerAccumulator.updateNumericalValue

The first `if` statement in the current implementation is redundant, and can be deleted:

```js
    updateNumericalValue: function( accumulatedKeys, index ) {
      if ( accumulatedKeys.length === 0 ) {
        return 0;
      }

      var stringRepresentation = this.updateStringValue( accumulatedKeys, index );
      return stringRepresentation.length > 0 ? parseInt( stringRepresentation, 10 ) : 0;
    },
```

**Comment by pixelzoom** [comment]
- [ ] IntegerAccumulator.validateAndProcessInput is really bad.

This is just a mess:

```js
      var length = accumulatedKeys.length;
      var multiplier = 1;
      var maxLength = this.options.maxLength;
      var startIndex = 0;
      var startString = '';
      if ( length > 0 && accumulatedKeys[ 0 ] instanceof PlusMinusKey ) {
        multiplier = -1;
        maxLength += 1;
        startIndex = 1;
        startString = '-';
      }
```

I'm going to go ahead and change it.  Ask me if you have questions.

**Comment by pixelzoom** [comment]
- [ ] Leading zero should be overwritten.

This was a problem before I made the above changes.  If you enter a '0', it should be overwritten by subsequent digits.  The display should not look like this:

![screenshot_101](https://cloud.githubusercontent.com/assets/3046552/21406216/a262f208-c787-11e6-873b-4331f092c391.png)

**Comment by pixelzoom** [comment]
OK, I'm done with this review.  There are a lot of open issues, represented by check-box items in the above comments.  

The big issues are:

(1) The client shouldn't need to know about accumulator; Keypad (and its subtypes) should provide the complete API.
(2) AbstractKey should not have an associated `{number} value`, which is relevant only for DigitKey.
(3) Specification of layouts should be simplified/clarified.
(4) DecimalAccumulator would help to flush out type hierarchy issues.
(5) Leading zeros are not handled properly.

**Comment by pixelzoom** [comment]
In retrospect... A separate issue (something like "initial review") should have been created for this review, rather than putting this in the "design requirements" issue.  Let's keep that in mind for the next review.

**Comment by pixelzoom** [comment]
A couple of other things that occurred to me, after ruminating on the current implementation. These are (imo) both big architectural flaws.

---

- [ ] Where is the ability to customize the constraints on what is entered?

In https://github.com/phetsims/scenery-phet/issues/272 we discussed the importance on being able to customize the constraints on what can be entered using the keypad.  And that issue resulted in the addition of `options.validateKey` in NumberKeypad.  I see no evidence of any similar API in this new implementation.  In fact, the only validation is a maximum number of digits that is hardwired into `IntegerAccumulator.validateKeys`:

```js
65 if ( this.getNumberOfDigits( proposedKeys ) <= this.options.maxLength ) {
```

Validation should be encapsulated in one place, and it should be customizable.

---

- [ ] Accumulating keys involves misplaced responsibilities and broken encapsulation.

My first red flag related to this was in AbstractKey.handleKeyPress, whose signature is:

```js
39 handleKeyPressed: function( keyAccumulator )
```

If feels wrong that keys must know about the accumulator.  The key creates a "proposed" set of keys, based on the current accumulated keys.  And then that proposed set is what is validated.  Keys are also responsible for handling the "clear on next pressed feature" (which results in duplicated code, btw.) And keys know the structure of how the accumulator stores keys - e.g., the keys are manipulating arrays.  This all seems very wrong to me - the accumulator should make decisions based on which key is pressed, and how keys are accumulated should be private to the accumulator.

**Comment by pixelzoom** [comment]
Another note on the validation issue mentioned in https://github.com/phetsims/scenery-phet/issues/283#issuecomment-268678103...  The current implementation of IntegerAccumulator only allows you to constrain the number of digits entered.  As an example of why this isn't sufficient... Unit Rates needs to constrain the number of digits _and_ requires that the entry is a positive entry (i.e., `0` is not valid.)  Future sims may have other constraints, and trying to foresee all constraints is impossible, so Keypad must allow clients to specify their own constraint algorithm.

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/scenery-phet/issues/283#issuecomment-268389241, @jbphet asked for feedback from me and @jonathanolson.  I don't see any feedback from @jonathanolson.  So I'm wondering if we should defer the design phone call scheduled for this morning?

**Comment by pixelzoom** [comment]
@jbphet and I discussed status via Skype, and we decided to defer our meeting until @jonathanolson has provided feedback.  

Here's the feedback I'd like to hear from @jonathanolson:
- Design of the Accumulator: responsibilities, encapsulation, implementation ideas.
- Input constraints: Thoughts on the API for constraining valid input, allowing customization by the client, etc.
- Review of my feedback, to see if there's anything you disagree with.

**Comment by jonathanolson** [comment]
I reviewed independently first, and came to the same high-level conclusions as @pixelzoom. Review here starts with higher-level points and moves towards low-level things I noticed. Feel free to convert to a checklist if it helps.

I was anticipating a structure where there was an enumeration of values:
```js
KeypadValue = {
  0: ...
  1: ...
  // ...
  9: ...
  MINUS: ...
  DECIMAL: ...
};
```

and the accumulator stored the current values (similar to what is current):
```js
// @public {Property.<Array.<KeypadValue>>}
this.valuesProperty = new Property( [] );
```

where each key contained logic just based on the keypad values:
```js
getValuesAfterPress: function( values: {Array.<KeypadValue>} ) : {Array.<KeypadValue}
```

The current structure is similar, but keys themselves are used in place of the values requiring extra documentation like:
```js
// PlusMinusKey (if present) will be first key, and indicates that the number is negative
if ( keys.length > 0 && keys[ i ] instanceof PlusMinusKey ) {
  returnValue = PlusMinusKey.MINUS_CHAR;
  i++;
}
```

and each key is instead given the full accumulator (and the responsibility to have to handle accumulator responsibilites, like clearing on next keypress):
```js
if ( keyAccumulator.getClearOnNextKeyPress() ) {
  newArray = [];
  keyAccumulator.setClearOnNextKeyPress( false );
}
else {
  // actually do things
}
```
This would prevent bugs like the BackspaceKey not fully clearing everything on the next keypress with clearOnNextKeyPress is set (it only removes one character and removes the flag).

Ideally, this would be handled in the accumulator:
```js
onPress: function( key ) {
  // Handle clearing on key presses in one place
  if ( this.clearOnNextKeyPress ) {
    this.valuesProperty.value = [];
    this.clearOnNextKeyPress = false;
  }
  else {
    var proposedValues = key.getValuesAfterPress( this.valuesProperty.value );
    if ( this.validate( proposedValues ) ) {
      this.valuesProperty.value = proposedValues;
    }
  }
}
```

So that keys could have a slightly simpler implementation, e.g. for DigitKey:
```js
getValuesAfterPress: function( currentValues ) {
  return Keypad.removeLeadingZero( currentValues ).concat( [ this.value ] );
}
```
instead of the current:
```js
handleKeyPressed: function( keyAccumulator ) {
  var newArray;
  if ( keyAccumulator.getClearOnNextKeyPress() ) {
    newArray = [];
    keyAccumulator.setClearOnNextKeyPress( false );
  }
  else{
    newArray = _.clone( keyAccumulator.accumulatedKeysProperty.get() );
    if ( keyAccumulator.valueProperty.get() === 0 ){
      newArray.pop();
    }
  }
  newArray.push( this );
  return newArray;
}
```

or for PlusMinusKey:
```js
getValuesAfterPress: function( currentValues ) {
  var isNegative = currentValues.length > 0 && currentValues[ 0 ] === KeypadValue.MINUS;

  // Remove minus sign if the value is already negative
  if ( isNegative ) {
    return currentValues.slice( 1 );
  }
  // Otherwise add a minus sign on the front
  else {
    return [ KeypadValue.MINUS ].concat( currentValues );
  }
}
```

Additionally, it would be ideal for the accumulator to have a custom validate() function given in the options (which
may include maxDigits handling). I agree with @pixelzoom on the need to have custom validation functions. Having a library (like Keypad.validateMaxDigits, Keypad.validateNonzero, etc.) that can be composed would be great, so you could:
```js
new Keypad( ..., function validate( values ) {
  // Allow up to 4 digits, don't allow 0 on its own, and don't allow the digit 5
  return Keypad.validateMaxDigits( values, 4 ) &&
         Keypad.validateNonzero( values ) &&
         !_.includes( values, KeypadValue[ 5 ] );
} );
```

I didn't see anything about clearOnNextKeyPress in the AbstractKeyAccumulator, but it is assumed by all keys (which only require an AbstractKeyAccumulator). Can this move to the AbstractKeyAccumulator?

I like having the currentValues be a property (or ObservableArray as @pixelzoom notes), since the stringValue and numberValue can be DerivedProperties. However, I'd have the numberValue created directly from the values, so we don't have to handle stripping out Unicode minus signs or other display complexities. Thus I'd like to see valuesToString( values ) and valuesToNumber( values ), and:
```js
// in Keypad
this.stringProperty = new DerivedProperty( [ accumulator.valuesProperty ], accumulator.valuesToString.bind( accumulator ) );
this.numberProperty = new DerivedProperty( [ accumulator.valuesProperty ], accumulator.valuesToNumber.bind( accumulator ) );
```

Documentation of layout (and possibly simplification of that as noted by @pixelzoom) would help a lot. Took a bit to figure out it was ```{Array.<Array.<AbstractKey|null>>}```. Also, it should be further simplified, as I think rows like ```[ new DigitKey( 5, { horizontalSpan: 5 } ), null, new DigitKey( 3 ) ]``` will improperly position the 3 key. There seem to be a lot of edge cases, so accumulating row/column offsets (instead of computing for each value) in layout would be more error-proof.

For loops for rows/columns (or multiple dimensions handled in general) are more readable with names:
```js
for ( var row = 0, ... ) {
  for ( var column = 0, ... ) {
    doSomething( values[ row ][ column ], values[ row ][ column - 1 ] );
  }
}
```

There's a lot of inconsistent formatting, particularly with array brackets, e.g. ```layout[ numRows - 1][i]``` should be ```layout[ numRows - 1 ][ i ]```.

createKeyNode types should have `{AbstractKey}` for keys and `{AbstractKeyAccumulator}` for the accumulator.

Default buttonColor and buttonFont are specified in two places.

It's weird that displayNode can be a `{string}` or `{Node}`, although AbstractKey only allows Node. Can we remove the ability to have strings, or fully support it? If we keep the string support, I'm worried about the asymmetric auto-scaling to a fixed dimension.

Noting that stringToInteger can return null is quite important.

**Comment by jonathanolson** [comment]
Also I want to note that having an enumeration for values isn't a strict requirement, but the abstraction sounds helpful in general. It may be overgeneralizing, and keys as values may be fine.

In the future, what happens if you have a key that conditionally wants to store one of a few values? It probably won't come up, but having a concrete list of what values can be in the array may just help comprehension.

Either way, thinks like isDigit() is probably a helpful utility function (currently done with instanceof DigitKey).

**Comment by pixelzoom** [comment]
Self-assigning to review for discussion on 2/20/17 @ 10am.

**Comment by jbphet** [comment]
Over the last few days I've discussed the design of the keypad with @pixelzoom, @jonathanolson, and @aadish .  Below is a summary of the decisions made based on those discussions.

On the topic of the Keypad API, we agreed that the client shouldn't need to know about accumulator. The Keypad (and its subtypes) should provide a complete API.  The fields should be accessible like this:
+ `keypad.stringProperty.link`
+ `keypad.valueProperty.link`
+ `keypad.accumulatedKeysProperty.link`

...versus something like this:
+ `keypad.accumulator.stringProperty.link`

Since we may sometimes have a need for custom accumulators, the keypads should have default accumulators with the ability to provide a different one.

On the topic of custom validation, we need to support this, so we decided to use a combination of general options so that clients don't have to create a validator any time they want any sort of variation, but also should allow the client to supply a completely different validator.  We are thinking that the API will offer options like this:

+ `maxDigitsLeftOfMantissa`
+ `maxDigitsRightOfMantissa`

...but also these:

+ `additionalValidator`
+ `alternativeValidator`

`additionalValidator` will be a function that was just in addition to the default validation.  `alternativeValidator` will be a different function and, if provided, would override all default and additional validation.

To make it easier to create new keypads, we could have a library of parameterized validators, creating them something like this:
```js
   options.additionalValidation = Keypad.currencyValidator( { maxDigitsLeftOfMantissa: 4 } )
```

We should be sure to note in the documentation that the validator can have side effects. Write in use case where decimal point is added.

On the topic of whether the keys should encapsulate the knowledge of how they are accumulated, we decided that the accumulator should have that knowledge.  This would allow keys to have different behaviors in different keypads.  For example, we may want the +/- key to work differently in number accumulator versus an equation accumulator.  The tradeoff is that we will need to change more files to add a new key.

We discussed using an observable array for the accumulated keys instead of a `Propery<Array>`, but decided that the latter will work better.

@pixelzoom suggested that we implement a floating point accumulator, determine what is common between this and the integer accumulator and consider putting it into a common base class.  @aadish will do this.

@pixelzoom and @jonathanolson - please comment if you think I've left anything out or misrepresented any of the things that we discussed.

@aadish is going to move forward with these changes and we will re-review once the main elements are working.

**Comment by pixelzoom** [comment]
@jbphet What's the status of the new keypad?

@ariel-phet FYI, the lack of closure on this issue is resulting in work-around solutions that are being copy-pasted between sims.  See https://github.com/phetsims/projectile-motion/issues/118

**Comment by jonathanolson** [comment]
> The fields should be accessible like this:
> 
> keypad.stringProperty.link
> keypad.valueProperty.link
> keypad.accumulatedKeysProperty.link

AbstractKeyAccumulator doesn't have stringProperty/valueProperty, that seems like an implementation detail in NumberAccumulator? (It's also not marked `@override` in NumberAccumulator).

TermAccumulator (in Area Model) has richStringProperty and termProperty (which somewhat map over), but initially there was no stringProperty equivalent (where I used `termProperty.value.toRichString()`) before I needed to handle the display slightly differently than the "value". I don't think assuming there's always a stringProperty equivalent is a safe assumption.

Are we sure we want to put restrictions on accumulators by pushing the client to forward things through Keypad itself?

> We should be sure to note in the documentation that the validator can have side effects.

That seems like an anti-pattern. Presumably validating would have no side effects?

As you refactor, please keep Area Model in mind (3 custom keypad layouts, custom accumulator).

**Comment by jonathanolson** [comment]
Also, it looks like these are already being forwarded (but some are undefined for Area Model's accumulator):

```js
    this.stringProperty = this.keyAccumulator.stringProperty; // @public (read-only)
    this.valueProperty = this.keyAccumulator.valueProperty; // @public (read-only)
    this.accumulatedKeysProperty = this.keyAccumulator.accumulatedKeysProperty; // @public (read-only)
```

Documentation for what the accumulator API is should probably be in the abstract accumulator, as if these were part of the API, it was undocumented.

**Comment by jbphet** [comment]
@jonathanolson said:

> AbstractKeyAccumulator doesn't have stringProperty/valueProperty, that seems like an implementation detail in NumberAccumulator?

and

>Are we sure we want to put restrictions on accumulators by pushing the client to forward things through Keypad itself?

This is a good point, and in fact, that's why I designed it this way in the first place - there will be different accumulators needed in different circumstances.  This makes it difficult to make these properties available directly from the keypad.  In a previous comment about design discussions with @pixelzoom and @jonathanolson I said, "we agreed that the client shouldn't need to know about accumulator".  I'd like to reverse this decision - it's a lot easier to implement and use this if the client knows about and hooks up to the accumulator.  It also makes sense to me - you have a keypad, and the keypad has a place where the keys get collected.  I think there will be enough examples available in the code so that new users will be able to make sense of this pretty quickly.

On the topic of validators, @pixelzoom said:

> We should be sure to note in the documentation that the validator can have side effects.

..and @jonathanolson said:

> That seems like an anti-pattern. Presumably validating would have no side effects?

I just spent some time looking at this, and I think it is no longer the case.  It may have been at one time - I honestly don't remember - but `AbstractKeyAccumulator.validateAndUpdate` is just expecting a boolean value from the validators.  I suppose there isn't anything that actively prevents the validators from modifying the proposed keys, but the default validator in `NumberAccumulator` doesn't do it, and there aren't any examples in the demo that do it either.  So, in short, I think this is no longer an issue.  Feel free to let me know if I've missed anything.

**Comment by jbphet** [comment]
I think the issues raised above have been addressed at this point, but there were a lot of them, so I have only a moderate degree of confidence that this is true.  I'm going to give @andrea-phet the go-ahead to use this in Projectile Motion, see https://github.com/phetsims/projectile-motion/issues/127.  I think the API is unlikely to change much at this point, so it should be safe.  I'm going to assign to @jonathanolson to comment in case he thinks it's *not* ready for this, since he has used it in Area Model.  If he's okay with it and doesn't veto, it should be assigned to @pixelzoom for another look when he returns.  If all looks good to him, it can be migrated into unit rates.  If not, I'll work with him to iron out any issues.

**Comment by jbphet** [comment]
Scanning back through the list of comments, I realized that I missed an item: implement dispose.  I'll do this shortly, but I think any review can occur in parallel.

**Comment by jonathanolson** [comment]
> I'm going to assign to @jonathanolson to comment in case he thinks it's not ready for this, since he has used it in Area Model. If he's okay with it and doesn't veto, it should be assigned to @pixelzoom for another look when he returns.

Sounds ready for projectile-motion usage to me. Assigning as described.

**Comment by pixelzoom** [comment]
@jbphet NumberAccumulator does not propagate options to AbstractKeyAccumulator, looks like a bug. I'm also guessing that the `alternativeValidator` and `additionalValidator` features of AbstractKeyAccumulator are therefore untested.

**Comment by pixelzoom** [comment]
@jbphet said:
> Scanning back through the list of comments, I realized that I missed an item: implement dispose. I'll do this shortly, but I think any review can occur in parallel.

Integration of Keypad into projectile-motion is completed.  Publication of that sim is blocked until `dispose` is implemented and `options` propagation is fixed (see https://github.com/phetsims/scenery-phet/issues/283#issuecomment-321404395).  High priority.

**Comment by ariel-phet** [comment]
@jonathanolson said he might be able to take care of this quickly. Since it is a blocking issue for projectile motion I ask him to move forward.

**Comment by jonathanolson** [comment]
As a side note, things would be more convenient if we had an array of validators, instead of having to branch to check for default/alternative/additional. It would also allow greater flexibility in the future.

**Comment by jonathanolson** [comment]
Added the options pass-through.

It seems like the only thing that needs disposal are the key nodes (as if the key labels are Nodes themselves, like in Area Model's case, they need the parent removed).

Am I missing something? (I double-checked all files in the keypad directory and really didn't see any complicated disposal that would be required).

**Comment by jbphet** [comment]
I added disposal of the properties that are created by the keypad.  I also added a test harness to the scenery-phet demo that removes one of the demo keypads and invokes the dispose function.

**Comment by jbphet** [comment]
@jonathanolson added the options pass through, and he and I added the dispose function.  However, I also added a test case to the scenery-phet demo that tests the disposal of the keypad, but that test case isn't passing yet.  The keypad is being successfully removed from the scene graph, and the dispose function is being called, and as far as I can tell all references from the demo are being cleaned up, yet the keypad is not being garbage collected.  I'll leave this open and investigate early next week.

**Comment by jbphet** [comment]
The issue that was causing the keypad to not be garbage collected has been resolved, and was unrelated to the dispose function.  It was an odd and subtle case, and a separate issue will be logged on it since @jonathanolson wanted to investigate it more after he and I looked at it.  I'm going to leave the test harness in the scenery-phet demo so that we can easily regression test the dispose function.  The test harness consists of a button that, when pressed, removes one of the keypads from the scene graph and calls its dispose function.  When using this, the idea is to load the keypad components, take a memory snapshot, then press the removal button and take another snapshot.  There should be three keypads in the first memory snapshot and two in the second.

This is what the keypad test/demo page looks like as of this writing, and the green button at the lower right is the one that tests the disposal:
![scenery-phet-test](https://user-images.githubusercontent.com/3082831/29288945-f7af7ebe-80f7-11e7-9432-879e6972571f.PNG)

At this point the options pass-through and dispose function should be taken care of.  Back to @pixelzoom for another check.

**Comment by pixelzoom** [comment]
Several things still to be addressed.

Little things:

- [ ]  Documentation in `NumberAccumulator` refers to non-existent `PlusMinusKey` in 2 places.

- [ ] `defaultValidator` should be `@protected` in both `AbstractKeyAccumulator` and `NumberAccumulator`. It’s intended to be overridden by subtypes, and shouldn’t be called by clients.

- [ ] `updateAndValidate` should be `@protected` in `AbstractKeyAccumulator`. It’s called by the subtype, not intended to be called by clients.

- [ ] There are several unused features in the implementation: `alternativeValidator`, `additionalValidator`, `KeyID.X`, `KeyID.X_SQUARED`, …  Have these features been tested?  If not, why not? And should ‘untested’ be noted in the internal documentation?

Big things:

- [ ] A string representation seems useful to (required by?) all accumulator subtypes. So why are `stringProperty` and `keyToString` specific to `NumberAccumulator`?  If that’s indeed true, then that functionality should be in `AbstractKeyAccumulator`.  (@jonathanolson  also commented on this on 7/26/17 in https://github.com/phetsims/scenery-phet/issues/283#issuecomment-318182334. I don’t see stringProperty addressed specifically in @jbphet ’s response https://github.com/phetsims/scenery-phet/issues/283#issuecomment-320366093.) And as mentioned in past comments, I think it’s difficult to get the accumulator hierarchy correct when there is only 1 concrete accumulator type (`NumberAccumulator`).  I’d really like to see another subtype, if only to flush out issues like the one preceding this one.

- [ ] @jonathanolson  said “Presumably validating would have no side effects?” JB said “I just spent some time looking at this, and I think it is no longer the case.”  Validation certainly does have side-effects, see `AbstractKeyAccumulator.validateAndUpdate`.  Validation of `proposedKeys` and update of `accumulatedKeysProperty` are tightly integrated into this one function.  If `proposedKeys` is rejected, it’s rejected silently, which seems like something that could be an issue in the future.  Highly recommended to decouple validation of `proposedKeys` and updating of `accumulatedKeysProperty`.

- [ ] In https://github.com/phetsims/scenery-phet/issues/283#issuecomment-321714316, JO suggested “things would be more convenient if we had an array of validators, instead of having to branch to check for default/alternative/additional. It would also allow greater flexibility in the future.”  I agree.  And I don’t see any comment that addresses this.

**Comment by jonathanolson** [comment]
> KeyID.X, KeyID.X_SQUARED, … Have these features been tested? If not, why not? And should ‘untested’ be noted in the internal documentation?

These are used in Area Model (see TermAccumulator). It was noted that any sim-specific Key IDs should be placed in KeyID, should that change?

>  I’d really like to see another subtype, if only to flush out issues of where responsibilities should reside.

TermAccumulator (not common code) is another subtype.

**Comment by jbphet** [comment]
Several of the "little things" have been addressed.

In the "big things" subcategory...

@pixelzoom said:

> A string representation seems useful to (required by?) all accumulator subtypes.

Maybe.  It just seemed to me that making any assumptions about what representations are needed makes for a brittle design, and I wanted to keep the abstract accumulator as simple and general as possible, so it only accumulates keys.  I could envision a case where only a numeric value is needed, for instance.  So, I'd prefer to keep it this way.  In the grand scheme of things, if there is a string representation in the base type, it probably won't ever mess anything up, so if others feel strongly I'll capitulate and move it there.

Also, @jonathanolson pointed out that there is a second accumulator sub-type - `TermAccumulator`.  @pixelzoom - do you feel that this sufficiently addresses the 2nd "Big Thing"?

I'll work on Little Thing 4 and Big Things 3 and 4 at a later time.  I'm not going to assign back @pixelzoom until I've done that, though he's welcome to comment when he sees this update if so inclined.

**Comment by pixelzoom** [comment]
Re questions in the above comment...

Perhaps @jonathanolson can chime in on whether he ran into hierarchy issue when creating `TermAccumulator`, and whether having `stringProperty` available in `AbstractKeyAccumulator` would have been useful.

**Comment by jonathanolson** [comment]
> Perhaps @jonathanolson can chime in on whether he ran into hierarchy issues when creating TermAccumulator, and whether having stringProperty available in AbstractKeyAccumulator would have been useful.

As noted in https://github.com/phetsims/scenery-phet/issues/283#issuecomment-318182334 above:

> TermAccumulator (in Area Model) has richStringProperty and termProperty (which somewhat map over), but initially there was no stringProperty equivalent (where I used termProperty.value.toRichString()) before I needed to handle the display slightly differently than the "value". I don't think assuming there's always a stringProperty equivalent is a safe assumption.

So having to require a stringProperty implementation for all accumulator subtypes doesn't "feel" correct to me.

The general accumulation and validation seemed quite general, like it should work for any subtype.

**Comment by pixelzoom** [comment]
Then I'm fine with considering "Big things" (1) and (2) non-issues.

**Comment by pixelzoom** [comment]
Labeling for developer meeting since https://github.com/phetsims/unit-rates/issues/207#issuecomment-434909102 reminded me that next-gen keypad is still not finished.

**Comment by pixelzoom** [comment]
Btw... This has been labeled with "meeting:design" since 11/8/2016.  Do designers not have a process for regularly reviewing labeled issues? If not, then why are we labeling them?

**Comment by pixelzoom** [comment]
Reopened, closed by mistake.

**Comment by jbphet** [comment]
In the 11/1/2018 design meeting it was decided to assign this issue to @Denz1994, since he may have some time as his current sim finishes up.  The goal would be to address the issues that are unchecked in the comments above.  Several of these are design issues, so this will require some ramping up on the overall design and the motives behind it.

**Comment by amanda-phet** [comment]
@pixelzoom said 
>Btw... This has been labeled with "meeting:design" since 11/8/2016 -- almost 2 years. Do designers not have a process for regularly reviewing labeled issues? If not, then why are we labeling them?

That is somewhat accurate. @arouinfar and I use the label within a sim repo to quickly find the issues we want to discuss in a design meeting. If someone uses the label, that person is responsible for holding a design meeting to discuss the labeled issue(s).

**Comment by Denz1994** [comment]
Addressing "little things" point (4) in https://github.com/phetsims/scenery-phet/issues/283#issuecomment-322519811:

`KeyID.X`, `KeyID.X_SQUARED` are used in AreaModelCommon.TermAccumulator.  Refactors were made by @jbphet 08/04/17 and that sim has gone through one round of dev testing (03/06/18) and one round of rc testing(04/24/18). Its current state seems stable does this suffice @pixelzoom?

**Comment by Denz1994** [comment]
Addressing "little things" point (4) in https://github.com/phetsims/scenery-phet/issues/283#issuecomment-322519811:

`AbstractKeyAccumalator.defaultValidator()` requires an override. 
```js
 /**
     * default validation, must be overridden in sub-types
     * @param {Array.<KeyID>} proposedKeys - the proposed set of keys to be validated
     * @returns {boolean}
     * @protected
     * @abstract
     */
    defaultValidator: function( proposedKeys ) {
      throw new Error( 'abstract function must be implemented by subtypes' );
    },
```
If the client always has to override the default validator this makes the option `alternativeValidator` redundant. The doc for alternativeValidator states `used instead of the default validation`. If we are always overriding the default validation(), that override is our “alternativeValidatior”.  Thoughts on removing `alternativeValidator` and keeping the override for `defaultValidator` @pixelzoom?

**Comment by Denz1994** [comment]
Addressing "big things" (3) in https://github.com/phetsims/scenery-phet/issues/283#issuecomment-322519811:

I refactored `AbstractKeyAccumulator.validateAndUpdate()` into `....validate()` and `....update()`. This was only being used in `AreaModelCommon.TermAccumulator` and in the ComponentScreenView of scenery-phet demo. I made the refactors at the callsite.

I attempted adding an assert when failing validate(proposedKeys) fails but that breaks the sim behavior everytime a key is pressed that may surpass the `maxDigits` allowed in the readout. It seems like a silent failure is what we want in this case. Thoughts @pixelzoom or is this a design question?

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/scenery-phet/issues/283#issuecomment-460806764, @Denz1994 asked:
> Addressing "little things" point (4) ... does this suffice @pixelzoom?

Good to hear that `KeyID.X` and `KeyID.X_SQUARED` are being used.  But this what about the other unused features in (4)?  

In https://github.com/phetsims/scenery-phet/issues/283#issuecomment-460806764, @Denz1994 asked:
> Addressing "little things" point (4) ... Thoughts on removing `alternativeValidator` and keeping the override for `defaultValidator` @pixelzoom?

Yes, I think it makes sense to delete `alternativeValidator`. AbstractKeyAccumulator requires a validator, and that can be done either by implementing an abstract method (the current approach) our adding a required `validator` constructor param.  If the validator needs to access  AbstractKeyAccumulator fields, then the former is probably the cleaner approach.  

In https://github.com/phetsims/scenery-phet/issues/283#issuecomment-460807305, @Denz1994 asked:
> Addressing "big things" (3) ... Thoughts @pixelzoom or is this a design question?

I'm not sure whether silently ignoring keys is a design or implementation issue. It's been so long since I did this review that I don't recall the tradeoffs.  Perhaps consult with @jbphet.

But there's a problem with this new implementation. If options documentation is correct, then AbstractKeyAccumulator `validateKeys` appears to have bad logic.  Option `additionalValidator` is supposed to be called _in addition to_ other validation, not ignored if `alternativeValidator` is provided. (It's also not a good practice to have multiple `return` statements in a method.)  So I would expect `validateKeys` to look like this:

```js
    validateKeys: function( proposedKeys ) {

      let valid = true;

      // alternativeValidator overrides defaultValidator
      if ( this.alternativeValidator ) {
        valid = this.alternativeValidator( proposedKeys );
      }
      else {
        valid = this.defaultValidator( proposedKeys );
      }

      // If additional validation is provided, do it here.
      if ( valid && this.additionalValidator ) {
        valid = this.defaultValidator( proposedKeys ) && this.additionalValidator( proposedKeys );
      }

      return valid;
    },
```

**Comment by pixelzoom** [comment]
And if `alternativeValidator` goes away:

```js
    validateKeys: function( proposedKeys ) {

      // default validation
      let valid = this.defaultValidator( proposedKeys );

      // .. followed by additional (optional) validation.
      if ( valid && this.additionalValidator ) {
        valid = this.additionalValidator( proposedKeys );
      }

      return valid;
    },
```

**Comment by Denz1994** [comment]
Addressing "big things" (4) and "little things" (4) from [#283 (comment)](https://github.com/phetsims/scenery-phet/issues/283#issuecomment-322519811)

It was suggested by @jonathanolson to use an array of functions instead of flags for validators. I really like this [suggestion](https://github.com/phetsims/scenery-phet/issues/283#issuecomment-321714316) as it removes the need for `alternativeValidator`, `defaultValidator`, and `additionalValidator`. Instead, this has been replaced with `validators`.

`AbstractKeyAccumulator`'s signiture has been changed to read:
`AbstractKeyAccumulator( validators, options )` where validators is of type `{Array.}`

The validation now looks like this:
```
  validateKeys: function( proposedKeys ) {

      // Ensures that proposedKeys exist before validation
      var valid = !!proposedKeys;

      // If any validator returns false then the proposedKey is not valid
      this.validators.forEach( function( validator ) {
        valid = valid && validator( proposedKeys );
      } );
      return valid;
    },
```

I tested the use of additional validators by adding the code block below in `TermAccumulator.js` and similarly in `NumberAccumulator.js`. This falsified validation and caused it to fail, which is good for the case below.

```
    // Testing an additional validator.
    validators.push(
      function() {
        return false;
      } );
```

@jonathanolson Could you please provide feedback if this satisfies what you had in mind?

**Comment by Denz1994** [comment]
Addressing "big things" (3) from https://github.com/phetsims/scenery-phet/issues/283#issuecomment-460807305:

>  If proposedKeys is rejected, it’s rejected silently, which seems like something that could be an issue in the future.

If the user puts in an invalid value wouldn't we want the `proposedKeys` to be silently rejected. The behavior currently follows:

`user clicks a key -> it is registered as invalid -> nothing happens `

IMO, this is the expected behavior of a keypad, unless we want specialized negative feedback for invalid values. Thoughts on this implementation @jbphet?

Also tagging @amanda-phet because I don't see any notes on this in the HTML5 NumberKeyPad design doc and @amanda-phet was the last designer to respond in this issue.

**Comment by pixelzoom** [comment]
@Denz1994 said:
> Addressing "big things" (3) from https://github.com/phetsims/scenery-phet/issues/283#issuecomment-460807305:

Seems like the wrong comment link. Did you mean https://github.com/phetsims/scenery-phet/issues/283#issuecomment-322519811?

**Comment by Denz1994** [comment]
@pixelzoom Correct sorry for the confusion. The correct comment is the one you [mentioned](https://github.com/phetsims/scenery-phet/issues/283#issuecomment-322519811).

**Comment by pixelzoom** [comment]
Thanks @Denz1994.  It's been awhile since my review, so I'm a little fuzzy on this issue.  But whether or not `proposedKeys` is silently rejected is not the primary point of that checklist item.  The primary point is the decoupling of `proposedKeys` and updating of `accumulatedKeysProperty` in `AbstractKeyAccumulator.validateAndUpdate`.  And since that method is currently commented out, I'm guessing that has been done or is in progress.

**Comment by Denz1994** [comment]
@pixelzoom Thanks for the feedback, the decoupling has been handled via `AbstractKeyAccumulator.validateKeys` and `AbstractKeyAccumulator.updateKeys`. Since the silent rejection isn't the main, issue I'll mark that item as completed and remove the relevant assignees.

**Comment by Denz1994** [comment]
The remaining unchecked item for this issue requires a brief review of the validator array implementation explained [here](https://github.com/phetsims/scenery-phet/issues/283#issuecomment-475034102).

**Comment by Denz1994** [comment]
Updated validator implementation has been reviewed by @jonathanolson as explained [here](https://github.com/phetsims/scenery-phet/issues/283#issuecomment-475034102).
Removing his assignment and proceeding to related issues.

**Comment by jonathanolson** [comment]
Eek, defaultValidator was being overridden in TermAccumulator, but it now doesn't exist in the base type!

**Comment by pixelzoom** [comment]
PhET still needs a keypad and a framework for popping up a keypad when clicking on a NumberDisplay.  This work has not been completed, and has been in progress since 2016.  

Fourier needs a keypad.  I can make Keypad work, but it still has issues, and would be better if it were finished.  There is no common-code support for popping up a Keypad, so I'll need to either make another copy of KeypadLayer (which originated in unit-rates and has already been copied by projectile-motion, see https://github.com/phetsims/unit-rates/issues/153) or copy KeypadDialog (a different approach that BL created for collision-lab, see https://github.com/phetsims/collision-lab/issues/60).  Multiple approaches is undesirable.

Assigning to @ariel-phet to evaluate priority.

**Comment by pixelzoom** [comment]
FYI... I've completed implementation of the keypad feature for Fourier.  It uses Keypad.js (the next-gen keypad), and a simpler version of the KeypadDialog.js approach that was used in collision-lab.

**Comment by ariel-phet** [comment]
With the understanding that a copy in Fourier will work, but add to technical debt, I understand that multiple approaches are undesirable, but this work is not currently a priority. Marking as deferred.

**Comment by zepumph** [comment]
From today's dev meeting, I may be able bring this home while working on Projectile Motion for PhET-iO.

**Comment by zepumph** [comment]
Discussed with @jbphet today, and we checked off the top checkboxes that have been tackled. 


Game plan for Keypad work:

-  [ ] go through this issue and clean up anything still to do
- [ ] Close this issue (once making sure all work is done or captured in side issues)
- [ ] https://github.com/phetsims/scenery-phet/issues/272
- [ ] https://github.com/phetsims/scenery-phet/issues/271 and https://github.com/phetsims/scenery-phet/issues/663
- [ ] some other issues I see around to work on during/after the above prioritized list:
  - [ ]  https://github.com/phetsims/scenery-phet/issues/540
  - [ ] https://github.com/phetsims/scenery-phet/issues/541
  - [ ] ~https://github.com/phetsims/scenery-phet/issues/333~ this is not part of the scope of work I plan to tackle 

Note I gathered this list from looking at https://github.com/issues?q=is%3Aopen+is%3Aissue+user%3Aphetsims+keypad as well as any linked issues around this issue. Anyone feel free to comment on loose ends or other thoughts I may not be thinking about.

**Comment by jbphet** [comment]
@zepumph is on the case, I'm going to unassign myself for now.

**Comment by markgammon** [comment]
Reopening and turning to an epic

**Comment by marlitas** [comment]
From a conversation with @kathy-phet this is not a priority for PhET-iO until a client asks for it. Completing keypad work is currently dependent on developer priority.

**Comment by jbphet** [comment]
The dev team has decided to make this somewhat of a priority, so I've used the "Monday request form" to request that this work be scheduled for an upcoming iteration.  I'm going to defer the issue until the work is scheduled.

**Comment by jbphet** [comment]
@zepumph and I went through the unchecked items on the list today, and he's going to take this over for a while.  It's not immediately clear to me which items are higher priority than the others, but the current context is that we need to start adding support for tier 1 description, and the first sim to use this will be My Solar System.  @zepumph will work from that perspective to prioritize and start on this list.

**Comment by zepumph** [comment]
I made a bit of progress here, and updated some of the issues. I focused on making sure all issues were linked in the initial comment here. I now need to move on to other priorities.
